registry/tests/api/main.rs

```diff
@@ -23,35 +29,68 @@ pub fn create_publish_request_sample() -> PublishRequest {
     }
 }
 
-#[tokio::test]
-async fn test_publish_registry() -> Result<(), Box<dyn std::error::Error>> {
+pub async fn basic_setup() -> RegistryClient<Channel> {
+    let (client, server) = tokio::io::duplex(1024);
+
     let path = Path::new("/tmp");
     let storage = Arc::new(storage::Filesystem::new(path));
     let metadata = Arc::new(InMemoryMetadataStorage::new());
-    let socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 0)), 1111);
-    let fake_server = Context::new(storage, metadata, socket);
+
+    let url = "0.0.0.0:0";
+
+    // this is needs to be removed once Context got cleaned up
+    let socket = SocketAddr::from_str(url).expect("Shouldn't happen");
+    let context = Context::new(storage, metadata, socket);
+
+    tokio::spawn(async move {
+        Server::builder()
+            .add_service(RegistryServer::new(context))
+            .serve_with_incoming(tokio_stream::once(Ok::<_, std::io::Error>(server)))
+            .await
+    });
+
+    // Move client to an option so we can _move_ the inner value
+    // on the first attempt to connect. All other attempts will fail.
+    let mut client = Some(client);
+    let channel = Endpoint::try_from(format!("http://{}", url))
+        .expect("Shouldn't happen")
+        .connect_with_connector(service_fn(move |_: Uri| {
+            let client = client.take();
+
+            async move {
+                if let Some(client) = client {
+                    Ok(client)
+                } else {
+                    Err(std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        "Client already taken",
+                    ))
+                }
+            }
+        }))
+        .await
+        .expect("Shouldn't happen");
+
+    RegistryClient::new(channel)
+}
+
+#[tokio::test]
+async fn test_publish_registry() {
+    let mut client = basic_setup().await;
 
     // 1. Insert a package and expect it to be successful
     {
         let req = tonic::Request::new(create_publish_request_sample());
-        let _res = fake_server.publish(req).await?;
+        client.publish(req).await.expect("Shouldn't happen");
```

Comment:
nit: if you don't provide any meaningful content in the expect statement id argue that you can just use unwrap